### PR TITLE
db-factfinder - incorporate new census data

### PR DESCRIPTION
**Pull request by fvankrieken**
https://github.com/NYCPlanning/edm-overview/issues/956

### CODE REVIEW

**Comment by fvankrieken** [comment]
One outstanding issue on folder structure - currently we export with `geog` as part of the folder structure. Python functionality has options for "2010", "2010_to_2020" (converting 2010 data to estimates in 2020 geographies), "2020"... but we exclusively output in "2010_to_2020" at the moment, so these options don't really have meaning for the outputs. I asked TM and he's unsure

File: dcpy/connectors/postgres.py [hunk 1]
```diff
@@ -3,6 +3,12 @@
 import csv
```
Comment (fvankrieken) [inline dcpy/connectors/postgres.py:3]: I figure `psql` generally refers to postgres command line utility, this is a more accurate name

File: dcpy/utils.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import subprocess
```
Comment (fvankrieken) [inline dcpy/utils.py:1]: should this be a folder maybe? Or either way is `utils` too generic? Could have this in a file/folder called `git` as well. 

About the actual contents - git username requires using git command line as far as I know. Not used in this codebase at the moment but is in data library. But for git_branch, should we maybe go with a different approach? Maybe in CI, use available environment variables rather than subprocess? Or either way, we could parse from `.git` folder. Curious on y'all's thoughts

**Comment by fvankrieken** [comment]
Not entirely related to this PR but just noticed - we have python-geosupport in requirements coming from a not-found commit hash. FacDB says this is supposed to be pinned to 21c. Any idea why that would be?

Dev container and build-geosupport images have latest geosupport (python included) installed, I think we should remove this in requirements and use latest for FacDB

File: .github/workflows/on_hold/factfinder_acs.yml
Comment (fvankrieken) [inline .github/workflows/on_hold/factfinder_acs.yml:1]: These 3 files are a little broken it seems - I still did a bit of updating in their just to keep things in line with our norms, but could be argued to just leave them untouched (see `db-factfinder/pipelines/acs.py`, `db-factfinder/pipelines/acs_community_profiles.py`, `db-factfinder/geolookup`. But I think moving workflows to some sort of folder like this makes sense - not sure when we're going to return to this if ever, so not quite "archive" because we might want it at some point.

File: db-facilities/facdb/cli.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 import typer
 
-from dcpy.connectors import psql
```
Comment (fvankrieken) [inline db-facilities/facdb/cli.py:8]: See comment in dcpy

**Comment by fvankrieken** [comment]
So one remaining thing really is docs - a bit orthogonal to this PR but I'll add to the wiki now. Aside from copying in wiki from pff, I'll add
- explanation of different outputs of PFF (4 entirely distinct outputs rather than 1 primary)
- explanation of what the input data updates are (and why we get them)
- "manual" updates vs currently not-used api-based updates

Still waiting on some clarity around output folder structure, if we need to be labelling geom in the folder structure if we just do the same label for every dataset even when it doesn't quite seem relevant. But in the meantime, think it would be worth getting this into main and iterating later as needed

File: dcpy/utils.py (hunk 2 above)
Comment (alexrichey) [reply dcpy/utils.py:1]: I have a weak preference would be for a git.py file. Or just using a [git python package](https://gitpython.readthedocs.io/en/stable/intro.html)? But this seems fine. 

> CI, use available environment variables rather than subprocess

I like this idea

File: dcpy/connectors/postgres.py (hunk 1 above)
Comment (alexrichey) [reply dcpy/connectors/postgres.py:3]: Good call. I think I called named it `psql` before sqlalchemy crept it.

File: python/requirements.in [hunk 4]
```diff
@@ -1,23 +1,24 @@
...
 Jinja2==3.1.2
 mapclassify==2.5.0
 matplotlib==3.7.1
+openpyxl==3.1.2
 pandas==1.5.3
 psycopg2-binary==2.9.5
 pytest==7.2.2
 pytest-cov==4.0.0
 python-dotenv==0.20.0
-python-geosupport @ git+https://github.com/NYCPlanning/python-geosupport.git@3b8df06
```
Comment (alexrichey) [inline python/requirements.in:15]: @fvankrieken 
So we do need this for `facdb`, but you mentioned it was causing an issue? When I pip install, I just get the following `WARNING: Did not find branch or tag '3b8df06', assuming revision or ref.` which is what we'd expect. 

I'd just pinned it to this commit [here](https://github.com/NYCPlanning/python-geosupport/commit/3b8df06da5800ff38d0b75bc0b27181fbc919c3b). Maybe we just unpin it, and use the latest?

Comment (fvankrieken) [reply python/requirements.in:15]: Re our convo, I was wrong about the install, BUT we've decided to unpin (and just rely on whatevers installed in docker image) and see if FacDB builds

**Comment by fvankrieken** [comment]
@damonmcc this is ready to go once more

**Comment by fvankrieken** [comment]
> @damonmcc this is ready to go once more

Actually let's see a github action build succeed with my latest changes before you dive in

File: .github/workflows/factfinder_decennial_manual_update.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+name: Population Fact Finder - Build Decennial Census Manual Update
+
+on:
+  workflow_dispatch:
+  push:
```
Comment (damonmcc) [inline .github/workflows/factfinder_decennial_manual_update.yml:5]: looks like this would upload on every push, and maybe every push on every branch?

Comment (damonmcc) [reply .github/workflows/factfinder_decennial_manual_update.yml:5]: oh i didn't see this was from ur commit just now to test it nvm

Comment (fvankrieken) [reply .github/workflows/factfinder_decennial_manual_update.yml:5]: Yup sorry. Build looks good though! I'll get rid of this in a moment

Comment (fvankrieken) [reply .github/workflows/factfinder_decennial_manual_update.yml]: Fixed and good to go

File: dcpy/connectors/edm/publishing.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+BUCKET = "edm-publishing"
+
+
+def upload(
+    output: Path,
+    publishing_folder: str,
+    acl: str,
+    *,
+    s3_subpath: str = None,
```
Comment (alexrichey) [inline dcpy/connectors/edm/publishing.py:15]: Do we have an established enough folder convention on edm-publishing to just infer the `s3_subpath` from, say, the "name" of the release, then potentially a version of it? 

And for an example of what I mean by "name," in `edm-publishing/ db-pluto / 2020-02-23 / output` it would be  "2020-02-03". Then some releases every now and then have a subfolder (e.g. `edm-publishing/ db-cbbr / fvankrieken-patch-pluto-gha / latest / output`)

I think that, ideally, a user of this module would never have to think about S3 paths

**Review by damonmcc — APPROVED** [review]

File: dcpy/connectors/edm/publishing.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    latest: bool = True,
+    include_foldername: bool = False,
+    max_files: int = 20
+):
+    if s3_subpath is None:
+        ## potential TODO, change this to environment variable
+        s3_subpath = git_branch()
```
Comment (alexrichey) [inline dcpy/connectors/edm/publishing.py:22]: This feels misplaced here. Would it be possible to have the callers of this method figure out the path / git_branch, and just ensure that the `s3_subpath` isn't null? Would help us toward separation of concerns with edm and git

File: dcpy/connectors/edm/publishing.py (hunk 6 above)
Comment (fvankrieken) [reply dcpy/connectors/edm/publishing.py:15]: Yeah that's a good point, this really does need "version" (or "name" as you're referring to it) should be an argument, defaulting to current date is certainly not valid for all datasets. And I guess to your point, a "branch name" or something along those lines COULD just be a version, though often within dev branches it can still be useful to generate multiple versions, but even then current defaults don't necessarily fully meet needs (if "version" is just date, can't generate two outputs in the same day). 

Tying in to your below comment here as well - our most consistent norm with output folders now is `{dataset}/{branch}/{version}/...` so wanted to go with that as a sort of default.

PFF has some oddness in that there are sort of "subproducts", but I guess a simple way to handle that would be just have "publishing_folder" be `db-factfinder/acs` etc. But I do sort of like the norm of just a few arguments that are just single folder names rather than paths and follow some sort of default structure, and then some optional arg the one I have which just gives complete control over folder path

Comment (fvankrieken) [reply dcpy/connectors/edm/publishing.py:15]: Or actually for PFF, this is my simplified version of the folder structure - it used to have another that seems redundant

```python
    upload(
        output=file,
        publishing_folder=f"db-factfinder",
        version=str(date.today()),
        acl="public-read",
        s3_subpath=Path(git_branch()) / export_type / year,
        latest=latest,
    )
```

Comment (fvankrieken) [reply dcpy/connectors/edm/publishing.py:15]: So in my mind, we sort of have the more "absolute" norms of top-level publishing folder and bottom level version folder, but in between can be somewhat custom. Definitely open to simplifying it in some way. @damonmcc your thoughts on all this?

File: dcpy/connectors/edm/publishing.py (hunk 7 above)
Comment (damonmcc) [reply dcpy/connectors/edm/publishing.py:22]: agree it'd be great to separate S3 from git. so maybe no default value for `s3_subpath`?

File: dcpy/connectors/edm/publishing.py (hunk 6 above)
Comment (damonmcc) [reply dcpy/connectors/edm/publishing.py:15]: I don't think we have or should impose a folder convention right now, but the most common pattern does seem to be `{dataset}/{branch}/{version}/...`. 

per our discussion issue on output folder naming (https://github.com/NYCPlanning/data-engineering/issues/98) and chats we've had about "should we have to build on main for a production release?", I'm not sure we'll like having `branch` be such a high-level part of the folder convention forever

TLDR: I'm in favor of flexibility for now and deciding on a convention later

**Review by damonmcc — APPROVED** [review]

File: dcpy/connectors/edm/publishing.py (hunk 6 above)
Comment (fvankrieken) [reply dcpy/connectors/edm/publishing.py:15]: @alexrichey I think we'll go that route for now, leaving an optional subpath with no default just to make sure that fine-grained control is allowed while still keeping the more reliable/consistent "publishing folder" and "version" args
